propagation_ot.go

```diff
@@ -191,66 +169,39 @@ func (p *binaryPropagator) Join(
 	if !ok {
 		return nil, opentracing.ErrInvalidCarrier
 	}
-	// Handle the trace, span ids, and sampled status.
-	var traceID, propagatedSpanID int64
-	var sampledByte byte
 
-	if err := binary.Read(carrier, binary.BigEndian, &traceID); err != nil {
-		if err == io.EOF {
-			return nil, opentracing.ErrTraceNotFound
-		}
-		return nil, opentracing.ErrTraceCorrupted
-	}
-	if err := binary.Read(carrier, binary.BigEndian, &propagatedSpanID); err != nil {
+	// Read the length of marshalled binary. io.ReadAll isn't that performant
+	// since it keeps resizing the underlying buffer as it encounters more bytes
+	// to read. By reading the length, we can allocate a fixed sized buf and read
+	// the exact amount of bytes into it.
+	var len uint32
+	if err := binary.Read(carrier, binary.BigEndian, &len); err != nil {
```

Comment:
I might be misreading the comments, but writing the size of the marshalled Protobuf message is mainly a performance optimization to prevent allocating extra space. If you encode that size using the builtin proto function `EncodeVarint` that Ben linked, the returned `[]byte` has a varying size. Check out this line https://github.com/gogo/protobuf/blob/200875106f3bf0eb01eb297dae30b250a25ffc84/proto/encode.go#L92.